core/execution/future/mark_price.go

```diff
@@ -0,0 +1,243 @@
+// Copyright (C) 2023 Gobalsky Labs Limited
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package future
+
+import (
+	"sort"
+	"time"
+
+	"code.vegaprotocol.io/vega/core/matching"
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/libs/num"
+	"code.vegaprotocol.io/vega/protos/vega"
+	snapshot "code.vegaprotocol.io/vega/protos/vega/snapshot/v1"
+)
+
+type CompositePriceCalculator struct {
+	config           *types.CompositePriceConfiguration
+	trades           []*types.Trade
+	sourceLastUpdate []int64
+	bookPriceAtTime  map[int64]*num.Uint
+	price            *num.Uint
+	// [0] trade mark price
+	// [1] book mark price
+	// [2] first oracel mark price
+	// [2+n] median mark price
+	priceSources []*num.Uint
+}
+
+const (
+	TradePriceIndex       = 0
+	BookPriceIndex        = 1
+	FirstOraclePriceIndex = 2
+)
+
+func NewCompositePriceCalculatorFromSnapshot(mp *num.Uint, mpc *snapshot.CompositePriceCalculator) *CompositePriceCalculator {
+	if mpc == nil {
+		// migration - for existing markets loaded from snapshot, set the configuration to default to use last trade price
+		// for mark price
+		return &CompositePriceCalculator{
+			config: &types.CompositePriceConfiguration{
+				DecayWeight:        num.DecimalZero(),
+				DecayPower:         num.DecimalZero(),
+				CashAmount:         num.UintZero(),
+				CompositePriceType: types.CompositePriceTypeByLastTrade,
+			},
+			trades:           []*types.Trade{},
+			price:            mp,
+			priceSources:     make([]*num.Uint, 1),
```

Comment:
nit-pick for sure, but given that we have `TradePriceIndex`, `BookPriceIndex` and `FirstOraclePriceIndex` constants, I think it'd make sense to `make([]*num.Uint, 3)` to ensure the constants will never be out of bounds